docs: adding guides for provider and examples (#80)
platform-mesh/platform-mesh.github.io: concepts/api-export-binding.md

:::

## The Three Objects

APIExport/APIBinding is built on three custom resources, each with a single responsibility.

| Object | Where it lives | Created by | Purpose |
|---|---|---|---|
| **APIResourceSchema** | Provider workspace | Provider (often generated from a CRD) | Defines a single API type. Structurally like a CRD, but creating it does not serve the API — the provider explicitly activates it via an APIExport. Immutable once created. |
| **APIExport** | Provider workspace | Provider | Publishes one or more APIResourceSchemas for consumption. References the schemas by name. Gets a unique identity assigned by kcp so multiple providers can export the same group/resource without collision. |
| **APIBinding** | Consumer workspace | Consumer | Binds the consumer's workspace to a provider's APIExport by path + name. Once bound, the exported APIs appear in `kubectl api-resources` for the consumer. |

::: tip apigen

> APIResourceSchemas are rarely written by hand. kcp ships an [`apigen` CLI](https://github.com/kcp-dev/sdk/tree/main/cmd/apigen) that generates them 1:1 from existing CRDs. Platform Mesh's [api-syncagent](/overview/api-syncagent) automates this further — it reads CRDs on the service cluster and creates the matching APIResourceSchemas in the provider workspace for you.

Reviewer: ```suggestion
APIResourceSchemas are rarely written by hand. kcp ships an [`apigen` CLI](https://github.com/kcp-dev/sdk/tree/main/cmd/apigen) that generates them 1:1 from existing CRDs. kcp's [api-syncagent](/overview/api-syncagent) automates this further — it reads CRDs on the service cluster and creates the matching APIResourceSchemas in the provider workspace for you.
```

Since api-syncagent is from kcp, or?

:::

## Binding Flow

The following sequence shows who does what — from API definition on the provider side to status reflection on the consumer side.

```mermaid
sequenceDiagram

>     participant Provider as Provider Workspace
>     participant kcp as kcp Control Plane
>     participant Consumer as Consumer Workspace
>     participant VW as Virtual Workspace
> 
>     Provider->>kcp: 1. Create APIResourceSchema
>     Provider->>kcp: 2. Create APIExport (references schema)
>     Note over kcp: kcp assigns identity,<br/>publishes virtual workspace endpoint
> 
>     Consumer->>kcp: 3. Create APIBinding (references export by path + name)
>     kcp->>kcp: 4. Validate binding, resolve identity
>     kcp->>Consumer: 5. APIs now available as native resources
> 
>     Consumer->>kcp: 6. Create resource (e.g., Widget)
>     Provider->>VW: 7. Watch resources via virtual workspace
>     VW->>Provider: 8. Creation event delivered
>     Provider->>Provider: 9. Reconcile (provision actual service)
>     Provider->>VW: 10. Update status
>     VW->>Consumer: Status reflected in consumer workspace

Reviewer: ```suggestion
    participant Provider as Provider Workspace
    participant kcp as kcp Control Plane
    participant Consumer as Consumer Workspace
    participant VW as Virtual Workspace

    Note over Provider,kcp: Initial setup by provider to expose APIs
    Provider->>kcp: 1. Create APIResourceSchema
    Provider->>kcp: 2. Create APIExport (references schema)
    Note over kcp: kcp assigns identity,<br/>publishes virtual workspace endpoint

    Note over kcp,Consumer: Consumer binds provier's API
    Consumer->>kcp: 3. Create APIBinding (references export by path + name)
    kcp->>kcp: 4. Validate binding, resolve identity
    kcp->>Consumer: 5. APIs now available as native resources

    Note over Provider,VW: Provider watches for resource events from Consumer
    Consumer->>kcp: 6. Create resource (e.g., Widget)
    Provider->>VW: 7. Watch resources via virtual workspace
    VW->>Provider: 8. Creation event delivered
    Provider->>Provider: 9. Reconcile (provision actual service)
    Provider->>VW: 10. Update status
    VW->>Consumer: Status reflected in consumer workspace
```

Reviewer: Move the comment from below as notes into the diagram, makes it easier to digest for new people I think.

```

> Steps 1-2 happen once during provider setup. Step 3 happens once per consumer binding. Steps 6-10 repeat for every resource lifecycle event.

Reviewer: ```suggestion
```

Claims are **two-sided**: the provider declares what it needs, and the consumer explicitly accepts each claim in their APIBinding. Neither side can unilaterally escalate. If the provider requests `["*"]` but the consumer accepts only `["get", "list"]`, the provider gets read-only access.

```yaml
# Provider declares (in APIExport):
spec:
  permissionClaims:
  - group: ""
    resource: configmaps
    verbs: ["get", "list", "create"]

# Consumer accepts (in APIBinding):
spec:
  permissionClaims:
  - resource: configmaps
    verbs: ["get", "list", "create"]
    state: Accepted
    selector:
      matchAll: true
```

Consumers can scope access further with `matchLabels` or `matchExpressions` selectors — restricting providers to a subset of objects in the workspace. New claims added to an existing APIExport do not retroactively apply to existing bindings; consumers must update the binding to accept them.

## Virtual Workspace (How Providers See Consumer Resources)

> Once a binding is active, the provider needs a way to watch and reconcile resources that consumers create — potentially across many consumer workspaces at once. kcp solves this with the **APIExport virtual workspace**: a proxy endpoint, published in the APIExport status, that aggregates all instances of the exported types across all bound consumers.

Reviewer: ```suggestion
Once a binding is active, the provider needs a way to watch and reconcile resources that consumers create — potentially across many consumer workspaces at once. kcp solves this with the **APIExport virtual workspace**: a proxy endpoint, published in the APIExportEndpointSlice status, that aggregates all instances of the exported types across all bound consumers.
```

APIExportES hasn't been discussed yet but here it should pivot to explaining what APIExportES are.

> Providers use the wildcard path (`clusters/*`) to watch every consumer workspace at once. Platform Mesh's [api-syncagent](/overview/api-syncagent) and controllers built with [multi-cluster-runtime](/overview/multi-cluster-runtime) both use this endpoint under the hood. You do not interact with the virtual workspace URL directly when you use either of those tools.

Reviewer: ```suggestion
Providers use the wildcard path (`clusters/*`) to watch every consumer workspace at once. Platform Mesh's [api-syncagent](/overview/api-syncagent) and controllers built with [multicluster-runtime](/overview/multi-cluster-runtime) both use this endpoint under the hood. You do not interact with the virtual workspace URL directly when you use either of those tools.
```

The upstream project is `multicluster-runtime`
